spring-cloud-data-rest/src/main/java/org/springframework/cloud/data/rest/controller/CounterController.java

```diff
@@ -0,0 +1,168 @@
+ */
+
+package org.springframework.cloud.data.rest.controller;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.actuate.endpoint.mvc.MetricsMvcEndpoint;
+import org.springframework.boot.actuate.metrics.Metric;
+import org.springframework.boot.actuate.metrics.repository.MetricRepository;
+import org.springframework.cloud.data.rest.resource.MetricResource;
+import org.springframework.cloud.data.rest.resource.CounterResource;
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.PageImpl;
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.web.PagedResourcesAssembler;
+import org.springframework.hateoas.ExposesResourceFor;
+import org.springframework.hateoas.PagedResources;
+import org.springframework.hateoas.ResourceAssembler;
+import org.springframework.hateoas.mvc.ResourceAssemblerSupport;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.MediaType;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestMethod;
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.ResponseBody;
+import org.springframework.web.bind.annotation.ResponseStatus;
+import org.springframework.web.bind.annotation.RestController;
+
+/**
+ * Allows interaction with Counters.
+ *
+ * @author Eric Bottard
+ */
+@RestController
+@RequestMapping("/metrics/counters")
+@ExposesResourceFor(CounterResource.class)
+public class CounterController {
+
+	public static final String COUNTER_PREFIX = "counter.";
+
+	@Autowired
+	private MetricRepository metricRepository;
+
+	private final ResourceAssembler<Metric<Double>, CounterResource> counterResourceAssembler =
+			new DeepCounterResourceAssembler();
+
+	protected final ResourceAssembler<Metric<Double>, ? extends MetricResource> shallowResourceAssembler =
+			new ShallowMetricResourceAssembler();
+
+	/**
+	 * List Counters that match the given criteria.
+	 */
+	@RequestMapping(value = "", method = RequestMethod.GET)
+	public PagedResources<? extends MetricResource> list(
+			Pageable pageable,
+			PagedResourcesAssembler<Metric<Double>> pagedAssembler,
+			@RequestParam(value = "detailed", defaultValue = "false") boolean detailed) {
+		/* Page */
+		Iterable metrics = metricRepository.findAll(/* pageable */);
+		Page<Metric<Double>> page = new PageImpl<>(filterCounters(metrics));
+		ResourceAssembler<Metric<Double>, ? extends MetricResource> assemblerToUse =
+				detailed ? counterResourceAssembler : shallowResourceAssembler;
+		return pagedAssembler.toResource(page, assemblerToUse);
+	}
+
+	/**
+	 * Retrieve information about a specific counter.
+	 */
+	@RequestMapping(value = "/{name}", method = RequestMethod.GET, produces = MediaType.APPLICATION_JSON_VALUE)
```

Comment:
Since we assume the controllers always produce `application/json` I thought we can remove `produces` attribute.